internal/n/crit_test.go

```diff
@@ -0,0 +1,190 @@
+package n_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/capatazlib/go-capataz/cap"
+	. "github.com/capatazlib/go-capataz/internal/stest"
+
+	"github.com/capatazlib/go-capataz/internal/n"
+	"github.com/capatazlib/go-capataz/internal/s"
+)
+
+var alwaysTrue n.EventCriteria = func(s.Event) bool { return true }
+var alwaysFalse n.EventCriteria = func(s.Event) bool { return false }
+
+func TestEAnd(t *testing.T) {
+	t.Run("all values are true", func(t *testing.T) {
+		counter := 0
+		evNotifier0 := func(s.Event) {
+			counter++
+		}
+		crit := n.EAnd(alwaysTrue, alwaysTrue, alwaysTrue)
+		evNotifier := n.SelectEventByCriteria(crit, evNotifier0)
+		evNotifier(s.Event{})
+		assert.Equal(t, 1, counter)
+	})
+
+	t.Run("one value is false", func(t *testing.T) {
+		counter := 0
+		evNotifier0 := func(s.Event) {
+			counter++
+		}
+		crit := n.EAnd(alwaysTrue, alwaysFalse, alwaysTrue)
+		evNotifier := n.SelectEventByCriteria(crit, evNotifier0)
+		evNotifier(s.Event{})
+		assert.Equal(t, 0, counter)
+	})
+}
+
+func TestEOr(t *testing.T) {
+	t.Run("all values are false", func(t *testing.T) {
+		counter := 0
+		evNotifier0 := func(s.Event) {
+			counter++
+		}
+		crit := n.EOr(alwaysFalse, alwaysFalse, alwaysFalse)
+		evNotifier := n.SelectEventByCriteria(crit, evNotifier0)
+		evNotifier(s.Event{})
+		assert.Equal(t, 0, counter)
+	})
+
+	t.Run("one value is true", func(t *testing.T) {
+		counter := 0
+		evNotifier0 := func(s.Event) {
+			counter++
+		}
+		crit := n.EOr(alwaysFalse, alwaysTrue, alwaysFalse)
+		evNotifier := n.SelectEventByCriteria(crit, evNotifier0)
+		evNotifier(s.Event{})
+		assert.Equal(t, 1, counter)
+	})
+}
+
+func TestENot(t *testing.T) {
+	crit1 := n.ENot(alwaysTrue)
+	crit2 := n.ENot(alwaysFalse)
+
+	counter := 0
+	evNotifier0 := func(s.Event) {
+		counter++
+	}
+
+	evNotifier1 := n.SelectEventByCriteria(crit1, evNotifier0)
+	evNotifier1(s.Event{})
+	assert.Equal(t, 0, counter)
+
+	evNotifier2 := n.SelectEventByCriteria(crit2, evNotifier0)
+	evNotifier2(s.Event{})
+	assert.Equal(t, 1, counter)
+}
+
+func TestEInSubtree(t *testing.T) {
+	counter := 0
+	evNotifier0 := func(s.Event) {
+		counter++
+	}
+
+	rootName := "root"
+	b0n := "branch0"
+	b1n := "branch1"
+
+	evNotifier := n.SelectEventByCriteria(n.EInSubtree(rootName, b0n), evNotifier0)
+
+	b0 := s.NewSupervisorSpec(b0n, s.WithNodes(WaitDoneWorker("child0")))
+	b1 := s.NewSupervisorSpec(b1n, s.WithNodes(WaitDoneWorker("child1")))
+
+	events, err := ObserveSupervisorWithNotifiers(
+		context.TODO(),
+		rootName,
+		s.WithNodes(
+			s.Subtree(b0),
+			s.Subtree(b1),
+		),
+		[]s.Opt{},
+		[]s.EventNotifier{evNotifier},
+		func(EventManager) {},
+	)
+
+	assert.NoError(t, err)
+
+	AssertExactMatch(t, events,
+		[]EventP{
+			// 1
+			WorkerStarted("root/branch0/child0"),
```

Comment:
I know this is just an internal test, but this is clearer to me: `"root/branch0"` than `n.EInSubtree(rootName, b0n)